### PR TITLE
Add ingestion layer for service discovery

### PR DESCRIPTION
**Pull request by hemantavi**
This PR adds support for the service discovery agent's ingestion layer.
These are the new packages that are added and their functionalities:
- mciutils: Has utility functions to parse/validate the MCI object. It
  also has the event handlers required for MCI informers.
- clusterset: Has utility functions to handle/validate clusterset
  object.
- svc_utils: Handles anything related to MCI services. It also has the
  filter created from the MCI services. It also has event handlers for
  member clusters' services and nodes.
- k8s_utils: Has functions to initialize a k8s member cluster. As a TODO,
  kubernetes resources like: services, nodes, pods, endpoints etc will
  be handled by this package.

Flow:
- The main function initializes the k8s module.
- The k8s module first fetches the clusterset object, the object is
  validated and k8s cluster objects are initialized from this object.
- The bootup handler is called, which fetches the MCI objects,
  validates them, fetches the services referred in the spec. With the
  fetched services, it creates a filter object.
- The bootup handler then fetches the services for each member cluster
  and applies the filter on this service. The bootup handler then
  proceeds to fetch the Node resources.
- After the bootup handler is done, the MCI informers are initialized
  and the service discovery agent then waits for MCI objects to be
  added.

This PR only contains the layer 1 functionality, so all queue related
functionalities/handling will be added as part of a subsequent PR.

Resolves AV-126257.

### CODE REVIEW

File: service_discovery/k8s_utils/event_handlers.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+	nodeEventHandler := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			node := obj.(*corev1.Node).DeepCopy()
+			gslbutils.Logf("cluster: %s, node: %s, msg: node added, will be published to layer 2",
+				c.Name(), node.GetName())
+		},
+		DeleteFunc: func(obj interface{}) {
+			node := obj.(*corev1.Node).DeepCopy()
+			gslbutils.Logf("cluster: %s, node: %s, msg: node added, will be published to layer 2",
```
Comment (monotosh-avi) [inline service_discovery/k8s_utils/event_handlers.go:76]: The log msg has to be changed for delete and update.

File: service_discovery/mci_utils/mci.go [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+	}
+	newSvcs, err := GetServiceList(newMCI)
+	if err != nil {
+		return svcToAdd, svcToDelete, fmt.Errorf("error in getting svc list from new MCI: %v", err)
+	}
+
+	// generate a map of both old and new services
+	oldSvcsMap := make(map[string]interface{})
+	for _, s := range oldSvcs {
+		oldSvcsMap[s.Cluster()+s.Namespace()+s.Name()] = struct{}{}
```
Comment (monotosh-avi) [inline service_discovery/mci_utils/mci.go:145]: Would it be better to add a string like "/" between cluster, namespace and name ? Else foo in namesapce bar would get mapped to same key as rfoo in namespace fo

Comment (hemantavi) [reply service_discovery/mci_utils/mci.go]: Yeah, there's a problem in the logic. Thanks! Fixed now.

File: service_discovery/k8s_utils/event_handlers.go (hunk 1 above)
Comment (hemantavi) [reply service_discovery/k8s_utils/event_handlers.go]: Thanks, fixed.

**Review by monotosh-avi — APPROVED** [review]

**Pull request merged**